File: problems/0020-valid-parentheses/solution.py
CLOSERS = {"(": ")", "[": "]", "{": "}"}


class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) % 2:
            return False
        stack: list[str] = []
        for c in s:
            if c in CLOSERS:
                stack.append(CLOSERS[c])
            elif not stack or c != stack.pop():
                return False
        return not stack

File: problems/0020-valid-parentheses/solution.ts
const closers: Record<string, string> = { "(": ")", "[": "]", "{": "}" };

function isValid(s: string): boolean {
  if (s.length % 2 !== 0) {
    return false;
  }
  const stack: string[] = [];
  for (const c of s) {
    const closer = closers[c];
    if (closer !== undefined) {
      stack.push(closer);
    } else if (c !== stack.pop()) {
      return false;
    }
  }
  return stack.length === 0;
}
